src/libstd/sys_common/gnu/libbacktrace.rs

```diff
@@ -121,12 +123,26 @@ pub fn print(w: &mut Write, idx: isize, addr: *mut libc::c_void,
     // case. There is no evidence at the moment to suggest that a more carefully
     // constructed file can't cause arbitrary code execution. As a result of all
     // of this, we don't hint libbacktrace with the path to the current process.
-    unsafe fn init_state() -> *mut backtrace_state {
+    //
+    // This can also return a File which should remain open until
+    // the first use of the state.
+    unsafe fn init_state() -> (*mut backtrace_state, Option<fs::File>) {
         static mut STATE: *mut backtrace_state = ptr::null_mut();
-        if !STATE.is_null() { return STATE }
-        STATE = backtrace_create_state(ptr::null(), 0, error_cb,
+        static mut FILENAME_BUF: *const Vec<i8> = ptr::null();
+        if !STATE.is_null() { return (STATE, None) }
+
+        let mut filename = ptr::null();
+        let mut file = None;
+        if let Ok(value) = ::sys::backtrace::get_executable_filename() {
```

Comment:
Nit 2: you can do `let (filename, file) = match ::sys::backtrace::get_executable_filename() { .... }`